File: src/parser/__tests__/fixtures/csharp/sample3.cs
// sample3.cs - Modern C# patterns: async/await, LINQ, extension methods, attributes

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace ModernPatterns
{
    // Extension methods
    public static class StringExtensions
    {
        public static bool IsNullOrEmpty(this string str)
        {
            return string.IsNullOrEmpty(str);
        }

        public static string Capitalize(this string str)
        {
            if (str.IsNullOrEmpty())
                return str;

            return char.ToUpper(str[0]) + str.Substring(1);
        }
    }

    public static class EnumerableExtensions
    {
        public static IEnumerable<T> WhereNotNull<T>(this IEnumerable<T> source)
        {
            return source.Where(item => item != null);
        }
    }

    // Attribute classes
    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method)]
    public class LogAttribute : Attribute
    {
        public string Message { get; set; }

        public LogAttribute(string message)
        {
            Message = message;
        }
    }

    [AttributeUsage(AttributeTargets.Property)]
    public class RequiredAttribute : Attribute
    {
    }

    // Async/await patterns
    public class AsyncDataService
    {
        public async Task<string> FetchDataAsync(string url)
        {
            await Task.Delay(100); // Simulate network delay
            return $"Data from {url}";
        }

        public async Task<T> FetchJsonAsync<T>(string url) where T : class
        {
            await Task.Delay(100);
            return default(T);
        }

        public async Task ProcessBatchAsync(IEnumerable<string> urls)
        {
            var tasks = urls.Select(url => FetchDataAsync(url));
            await Task.WhenAll(tasks);
        }
    }

    // Builder pattern
    public class QueryBuilder
    {
        private string _table;
        private List<string> _columns = new List<string>();
        private string _whereClause;
        private int? _limit;

        public QueryBuilder From(string table)
        {
            _table = table;
            return this;
        }

        public QueryBuilder Select(params string[] columns)
        {
            _columns.AddRange(columns);
            return this;
        }

        public QueryBuilder Where(string condition)
        {
            _whereClause = condition;
            return this;
        }

        public QueryBuilder Limit(int limit)
        {
            _limit = limit;
            return this;
        }

        public string Build()
        {
            var columns = _columns.Any() ? string.Join(", ", _columns) : "*";
            var query = $"SELECT {columns} FROM {_table}";

            if (!string.IsNullOrEmpty(_whereClause))
                query += $" WHERE {_whereClause}";

            if (_limit.HasValue)
                query += $" LIMIT {_limit.Value}";

            return query;
        }

        public static QueryBuilder Create()
        {
            return new QueryBuilder();
        }
    }

    // State machine pattern
    public enum MachineState
    {
        Idle,
        Running,
        Paused,
        Stopped
    }

    public class StateMachine
    {
        public MachineState CurrentState { get; private set; }

        public StateMachine()
        {
            CurrentState = MachineState.Idle;
        }

        public void Start()
        {
            if (CurrentState != MachineState.Idle && CurrentState != MachineState.Stopped)
                throw new InvalidOperationException($"Cannot start from {CurrentState}");

            CurrentState = MachineState.Running;
        }

        public void Pause()
        {
            if (CurrentState != MachineState.Running)
                throw new InvalidOperationException($"Cannot pause from {CurrentState}");

            CurrentState = MachineState.Paused;
        }

        public void Resume()
        {
            if (CurrentState != MachineState.Paused)
                throw new InvalidOperationException($"Cannot resume from {CurrentState}");

            CurrentState = MachineState.Running;
        }

        public void Stop()
        {
            CurrentState = MachineState.Stopped;
        }

        public void Reset()
        {
            CurrentState = MachineState.Idle;
        }
    }

    // Singleton pattern
    public sealed class Configuration
    {
        private static Configuration _instance;
        private static readonly object _lock = new object();

        public string Host { get; set; }
        public int Port { get; set; }
        public int Timeout { get; set; }

        private Configuration()
        {
            Host = "localhost";
            Port = 8080;
            Timeout = 30;
        }

        public static Configuration Instance
        {
            get
            {
                if (_instance == null)
                {
                    lock (_lock)
                    {
                        if (_instance == null)
                        {
                            _instance = new Configuration();
                        }
                    }
                }
                return _instance;
            }
        }

        public static void Reset()
        {
            lock (_lock)
            {
                _instance = null;
            }
        }
    }

    // Model base class
    public abstract class Model
    {
        public int Id { get; set; }
        public DateTime CreatedAt { get; set; }
        public DateTime UpdatedAt { get; set; }

        protected Model()
        {
            CreatedAt = DateTime.UtcNow;
            UpdatedAt = DateTime.UtcNow;
        }

        public virtual void Save()
        {
            UpdatedAt = DateTime.UtcNow;
            // Save logic
        }

        public virtual void Delete()
        {
            // Delete logic
        }
    }

    // User model
    [Log("User model accessed")]
    public class User : Model
    {
        [Required]
        public string Name { get; set; }

        [Required]
        public string Email { get; set; }

        public int Age { get; set; }

        public User()
        {
        }

        public User(string name, string email, int age)
        {
            Name = name;
            Email = email;
            Age = age;
        }

        public bool IsAdult()
        {
            return Age >= 18;
        }

        public static User FindByEmail(string email)
        {
            // Find user by email
            return null;
        }
    }

    // Post model
    public class Post : Model
    {
        public string Title { get; set; }
        public string Content { get; set; }
        public int UserId { get; set; }
        public bool Published { get; set; }

        public User GetAuthor()
        {
            // Get author by UserId
            return null;
        }

        public static IEnumerable<Post> GetPublished()
        {
            // Get all published posts
            return Enumerable.Empty<Post>();
        }
    }

    // Validation service
    public interface IValidator<T>
    {
        ValidationResult Validate(T item);
    }

    public class ValidationResult
    {
        public bool IsValid { get; set; }
        public List<string> Errors { get; set; }

        public ValidationResult()
        {
            Errors = new List<string>();
            IsValid = true;
        }
    }

    public class UserValidator : IValidator<User>
    {
        public ValidationResult Validate(User user)
        {
            var result = new ValidationResult();

            if (string.IsNullOrEmpty(user.Name))
            {
                result.IsValid = false;
                result.Errors.Add("Name is required");
            }

            if (string.IsNullOrEmpty(user.Email))
            {
                result.IsValid = false;
                result.Errors.Add("Email is required");
            }

            if (user.Age < 0)
            {
                result.IsValid = false;
                result.Errors.Add("Age must be positive");
            }

            return result;
        }
    }

    // Cache service
    public class CacheService<TKey, TValue>
    {
        private readonly Dictionary<TKey, CacheEntry<TValue>> _cache;
        private readonly TimeSpan _defaultTTL;

        public CacheService(TimeSpan defaultTTL)
        {
            _cache = new Dictionary<TKey, CacheEntry<TValue>>();
            _defaultTTL = defaultTTL;
        }

        public void Set(TKey key, TValue value)
        {
            _cache[key] = new CacheEntry<TValue>(value, DateTime.UtcNow.Add(_defaultTTL));
        }

        public bool TryGet(TKey key, out TValue value)
        {
            if (_cache.ContainsKey(key))
            {
                var entry = _cache[key];
                if (entry.ExpiresAt > DateTime.UtcNow)
                {
                    value = entry.Value;
                    return true;
                }
                else
                {
                    _cache.Remove(key);
                }
            }

            value = default(TValue);
            return false;
        }

        public void Clear()
        {
            _cache.Clear();
        }

        private class CacheEntry<T>
        {
            public T Value { get; set; }
            public DateTime ExpiresAt { get; set; }

            public CacheEntry(T value, DateTime expiresAt)
            {
                Value = value;
                ExpiresAt = expiresAt;
            }
        }
    }

    // Error classes
    public class ApplicationException : Exception
    {
        public ApplicationException(string message) : base(message)
        {
        }
    }

    public class ValidationException : ApplicationException
    {
        public ValidationException(string message) : base(message)
        {
        }
    }

    public class NotFoundException : ApplicationException
    {
        public NotFoundException(string message) : base(message)
        {
        }
    }

    // Constants
    public static class AppConstants
    {
        public const int MaxRetries = 3;
        public const int DefaultTimeout = 60;
        public const int BufferSize = 8192;
        public const string AppVersion = "3.0.0";
    }
}

File: src/parser/__tests__/fixtures/ruby/sample3.rb
# sample3.rb - Advanced Ruby patterns: metaprogramming, singleton methods, etc.

# Config class with singleton pattern
class Config
  @instance = nil

  attr_accessor :host, :port, :timeout

  def initialize
    @host = 'localhost'
    @port = 8080
    @timeout = 30
  end

  def self.instance
    @instance ||= new
  end

  def self.reset
    @instance = nil
  end

  private_class_method :new
end

# Builder pattern for configuration
class ConfigBuilder
  attr_reader :config

  def initialize
    @host = nil
    @port = nil
    @timeout = nil
  end

  def host(value)
    @host = value
    self
  end

  def port(value)
    @port = value
    self
  end

  def timeout(value)
    @timeout = value
    self
  end

  def build
    raise ArgumentError, "Host is required" if @host.nil?

    config = OpenStruct.new
    config.host = @host
    config.port = @port || 8080
    config.timeout = @timeout || 30
    config
  end

  def self.create
    new
  end
end

# State machine
class StateMachine
  attr_reader :current_state

  STATES = [:idle, :running, :paused, :stopped].freeze

  def initialize
    @current_state = :idle
  end

  def start
    raise "Cannot start from #{@current_state}" unless can_start?
    @current_state = :running
  end

  def pause
    raise "Cannot pause from #{@current_state}" unless @current_state == :running
    @current_state = :paused
  end

  def resume
    raise "Cannot resume from #{@current_state}" unless @current_state == :paused
    @current_state = :running
  end

  def stop
    @current_state = :stopped
  end

  def reset
    @current_state = :idle
  end

  private

  def can_start?
    [:idle, :stopped].include?(@current_state)
  end
end

# Active Record-like base class
class Model
  @@table_name = nil

  def self.table_name=(name)
    @@table_name = name
  end

  def self.table_name
    @@table_name || name.downcase + 's'
  end

  def self.all
    # Simulate database query
    []
  end

  def self.find(id)
    # Simulate database query
    nil
  end

  def self.where(conditions)
    # Simulate database query
    []
  end

  def save
    # Simulate saving to database
    true
  end

  def update(attributes)
    attributes.each do |key, value|
      send("#{key}=", value) if respond_to?("#{key}=")
    end
    save
  end

  def destroy
    # Simulate deleting from database
    true
  end
end

# User model
class User < Model
  self.table_name = 'users'

  attr_accessor :name, :email, :age

  def initialize(name: nil, email: nil, age: nil)
    @name = name
    @email = email
    @age = age
  end

  def full_name
    @name
  end

  def self.find_by_email(email)
    where(email: email).first
  end

  def self.adults
    where('age >= 18')
  end
end

# Post model
class Post < Model
  attr_accessor :title, :content, :user_id

  def initialize(title: nil, content: nil, user_id: nil)
    @title = title
    @content = content
    @user_id = user_id
  end

  def author
    User.find(@user_id)
  end

  def self.published
    where(published: true)
  end
end

# Validator module
module Validator
  def self.included(base)
    base.extend(ClassMethods)
    base.class_eval do
      @validations = []
    end
  end

  module ClassMethods
    def validates(field, options = {})
      @validations ||= []
      @validations << { field: field, options: options }
    end

    def validations
      @validations || []
    end
  end

  def valid?
    self.class.validations.all? do |validation|
      field = validation[:field]
      value = send(field)

      if validation[:options][:presence]
        !value.nil? && value != ''
      else
        true
      end
    end
  end

  def errors
    []
  end
end

# Product with validations
class Product
  include Validator

  attr_accessor :name, :price, :quantity

  validates :name, presence: true
  validates :price, presence: true
  validates :quantity, presence: true

  def initialize(name: nil, price: nil, quantity: nil)
    @name = name
    @price = price
    @quantity = quantity
  end

  def in_stock?
    @quantity > 0
  end

  def self.expensive(threshold)
    # Simulate query
    []
  end
end

# Cacheable module
module Cacheable
  def self.included(base)
    base.extend(ClassMethods)
  end

  module ClassMethods
    def cache_method(method_name)
      original_method = instance_method(method_name)

      define_method(method_name) do |*args|
        @cache ||= {}
        cache_key = [method_name, args].hash

        @cache[cache_key] ||= original_method.bind(self).call(*args)
      end
    end
  end

  def clear_cache
    @cache = {}
  end
end

# Calculator with caching
class CachedCalculator
  include Cacheable

  def fibonacci(n)
    return n if n <= 1
    fibonacci(n - 1) + fibonacci(n - 2)
  end

  cache_method :fibonacci

  def factorial(n)
    return 1 if n <= 1
    n * factorial(n - 1)
  end

  cache_method :factorial
end

# Constants
MAX_RETRIES = 3
DEFAULT_TIMEOUT = 60
BUFFER_SIZE = 8192
APP_VERSION = "3.0.0"

# Error classes
class ApplicationError < StandardError; end
class ValidationError < ApplicationError; end
class NotFoundError < ApplicationError; end

# Module-level helper methods
def retry_operation(max_attempts = MAX_RETRIES, &block)
  attempts = 0
  begin
    attempts += 1
    block.call
  rescue => e
    retry if attempts < max_attempts
    raise e
  end
end

def with_timeout(seconds = DEFAULT_TIMEOUT, &block)
  # Simulate timeout handling
  block.call
end
